### PR TITLE
Refactor MyTardis client for batched GET requests

### PR DESCRIPTION
**Pull request by andrew-uoa**
This PR is focused on the `mytardis_client` module. The main changes are:

- Add `get` and `get_all` methods to the MyTardis client class (`MyTardisRESTFactory`)
    - These have more strongly-typed interfaces than the general request method
    - These will support subsequent changes to speed up ingestion status-checking by batch-downloading all ingested datafiles under a dataset and doing local checking
    - In particular, `get_all` encapsulates the details of pagination in MyTardis responses. `get` is for the case where the requester only wants or expects a handful of objects (often just one)
- Add tests and associated fixtures for these new methods 
- Centralize more information about endpoints (supported HTTP methods, paths, response information) in the `endpoints` module in `mytardis_client`
    - This moves the remaining endpoint information from the top-level `helpers` module (now removed) into the MyTardis client
    - Part of the aim is to further clarify the object/endpoint distinction in the code, which has been blurred.

### CODE REVIEW

File: src/mytardis_client/endpoints.py [hunk 1]
```diff
@@ -1,60 +1,194 @@
 """Information about MyTardis endpoints"""
 
+from typing import Literal, Optional
+
 from pydantic import BaseModel
 
 from src.mytardis_client.objects import MyTardisObject
 
+MyTardisEndpoint = Literal[
```
Comment (JLoveUOA) [inline src/mytardis_client/endpoints.py:9]: Should this be only valid endpoints or only endpoints we want the ingestion scripts to be using?
if it's the former /group and /schema are both valid endpoints (and I might end up using /schema to validate RO-Crate metadata before ingestion)

Comment (andrew-uoa) [reply src/mytardis_client/endpoints.py:9]: Good point, it wasn't totally clear in my head. I think I assumed there were some I didn't know about, and would add as they came up, so I've added those, thanks

File: tests/test_mytardis_client_rest_factory.py [hunk 2]
```diff
@@ -70,5 +77,113 @@ def test_backoff_on_mytardis_rest_factory(
...
+        "/dataset_file",
+        object_type=Datafile,
+        batch_size=20,
+    )
+
+    assert isinstance(datafiles, list)
+    assert len(datafiles) == 30
+    assert (isinstance(df, Datafile) for df in datafiles)
+
+    assert datafiles[0].resource_uri == URI("/api/v1/dataset_file/0/")
```
Comment (JLoveUOA) [inline tests/test_mytardis_client_rest_factory.py:181]: could do the assert check in a loop here to check all 30 give you're confirming more than one already.

Comment (andrew-uoa) [reply tests/test_mytardis_client_rest_factory.py]: Good point, I've added that in. I was a bit reluctant as sometimes if you loop things and get an error, the message can be less explicit, so it's harder to figure out the issue. Will see how it goes though. In the C++ gtest framework, there's some nice stuff for testing that sequences of possibly different types are equal in some sense, I wonder if there's anything similar for python

**Review by JLoveUOA — APPROVED** [review]
Most issues can be addressed in next PR when integrating get_all() with observer, otherwise looks good.

File: src/mytardis_client/endpoints.py [hunk 3]
```diff
@@ -1,60 +1,196 @@
...
+
+    GET: Optional[GetRequestProperties] = None
+    POST: Optional[PostRequestProperties] = None
+
+
+class MyTardisEndpointInfo(BaseModel):
     """Properties of a MyTardis endpoint"""
 
-    url_suffix: str
+    path: str
```
[the remark this answers is not in the document]
Comment (andrew-uoa) [reply src/mytardis_client/endpoints.py]: Good spotting! I've made that change

File: src/mytardis_client/mt_rest.py [hunk 4]
```diff
@@ -139,3 +213,92 @@ def mytardis_api_request(
...
+                f"Endpoint: {endpoint}\n"
+                f"Query params: {query_params}\n"
+                f"Response: {response_json}"
+            )
+
+        if not isinstance(response_objects, list):
+            response_objects = [response_objects]
+
+        for object_json in response_objects:
+            obj = object_type.model_validate(object_json)
```
[the remark this answers is not in the document]
Comment (andrew-uoa) [reply src/mytardis_client/mt_rest.py:264]: Good question, no we didn't. We tended just to pluck values out of the JSON `dict` on an ad-hoc basis, where needed in the code. e.g. see `_get_uri_from_response()` in the Forge.

This has the benefit that we only reference the minimal set of fields needed to do the ingestion work, as opposed to setting out all the fields in a dataclass and doing a stricter validation, which might complain about missing fields that we don't actually need to use. Potentially this makes it easier to maintain compatibility with the API - fields can be added/removed from responses without change to the ingestion code, as long as they're not used anywhere. Maybe you'd call it looser coupling between the client and the server. On the other hand, in recent discussions we decided we want to bump up our component versions in lockstep with MyTardis, so maybe we _want_ tighter coupling here.

The downside is late validation - we don't detect the problem till we try to use the field, and it's less obvious in the context that the problem is a deficient API response. It also means more weakly typed interfaces in the ingestion code; more parameters that are just JSON blobs or dicts without much documentation of structure. It gets closer to the dreaded "stringly-typed" antipattern.

Another consideration is at some stage we might try to automatically generate the MyTardis client definition, based on the OpenAPI spec (need to upgrade the the API to use something that can produce OpenAPI spec first), so maybe we should take the current code closer to whatever that looks like.

e.g. we might use something like this:
https://github.com/koxudaxi/datamodel-code-generator/

[the remark this answers is not in the document]
Comment (andrew-uoa) [reply src/mytardis_client/mt_rest.py:264]: Good topic for a team discussion I reckon

**Pull request merged**